File: packages/onboarding/src/valuemaxx/onboarding/redact.py
"""Secret detection + redaction — the mechanical "no secret ever leaves" guarantee.

Every string the scanner captures from the user's codebase passes through
:func:`redact` before it can reach a proposal field, a diff hunk, or a log line
(design §7 / H12). Detection is conservative and additive: it anchors on the key
shapes this product encounters (Anthropic/OpenAI ``sk-`` keys, Anthropic admin
keys, AWS ``AKIA`` access-key ids, ``Bearer`` tokens), plus an assignment-form
heuristic (``API_KEY = "..."``) and a high-entropy-blob heuristic for credentials
that carry no recognisable prefix.

:func:`contains_secret` answers "is there a secret-shaped token here?";
:func:`redact` replaces every such token with :data:`REDACTION_PLACEHOLDER` and is
idempotent; :func:`assert_no_secret` raises :class:`SecretEncounteredError` when a
string that must be secret-free is not — it is the last gate before any output.
"""

from __future__ import annotations

import math
import re
from typing import Final

from valuemaxx.onboarding import rules
from valuemaxx.onboarding.errors import SecretEncounteredError

# Redaction constants come from the single-source ``rules`` module (serialized to the
# cross-language fixture the TS redactor reads), so Python + TS scrub the same shapes.
REDACTION_PLACEHOLDER: Final[str] = rules.REDACTION_PLACEHOLDER

# Prefixed key shapes. Anchored on known prefixes so ordinary identifiers (run
# ids, outcome names, status strings) are never scrubbed by accident. The admin
# pattern is listed first so it wins over the generic ``sk-ant`` pattern.
_PREFIX_PATTERNS: Final[tuple[re.Pattern[str], ...]] = tuple(
    re.compile(p) for p in rules.REDACT_PREFIX_PATTERNS
)

# Assignment form: a secret-named identifier set to a non-trivial value, e.g.
# ``API_KEY = "..."`` / ``password='...'`` / ``auth_token: ...``. Captures the
# value side (group ``val``) so we redact the secret, not the whole assignment.
_ASSIGNMENT_PATTERN: Final[re.Pattern[str]] = re.compile(
    rf"(?i)\b(?:{rules.REDACT_SECRET_NAME_ALT})\b\s*[:=]\s*"
    rf"(?P<val>(?:\"[^\"]{{6,}}\"|'[^']{{6,}}'|[^\s'\"]{{6,}}))",
)

# A long unbroken run of base64/credential-ish characters with no whitespace.
# Real prose and code break into short whitespace-separated tokens, so this fires
# only on credential-like blobs.
_HIGH_ENTROPY_TOKEN: Final[re.Pattern[str]] = re.compile(rules.REDACT_HIGH_ENTROPY_PATTERN)
_HIGH_ENTROPY_BITS: Final[float] = rules.REDACT_HIGH_ENTROPY_BITS


def _shannon_entropy(token: str) -> float:
    """Bits-per-character Shannon entropy of ``token`` (0.0 for an empty string)."""
    if not token:
        return 0.0
    counts: dict[str, int] = {}
    for char in token:
        counts[char] = counts.get(char, 0) + 1
    length = len(token)
    return -sum((c / length) * math.log2(c / length) for c in counts.values())


def _is_high_entropy_secret(token: str) -> bool:
    """True iff ``token`` is a long, high-entropy, multi-class credential blob."""
    if len(token) < 40:
        return False
    has_upper = any(c.isupper() for c in token)
    has_lower = any(c.islower() for c in token)
    has_digit = any(c.isdigit() for c in token)
    # Require mixed character classes — a long lowercase word is not a credential.
    if not (has_upper and has_lower and has_digit):
        return False
    return _shannon_entropy(token) >= _HIGH_ENTROPY_BITS


def _high_entropy_spans(text: str) -> list[tuple[int, int]]:
    """The (start, end) spans of high-entropy credential blobs in ``text``."""
    return [
        m.span() for m in _HIGH_ENTROPY_TOKEN.finditer(text) if _is_high_entropy_secret(m.group(0))
    ]


def contains_secret(text: str) -> bool:
    """Return True iff ``text`` contains a secret-shaped token."""
    if any(pattern.search(text) for pattern in _PREFIX_PATTERNS):
        return True
    if _ASSIGNMENT_PATTERN.search(text):
        return True
    return bool(_high_entropy_spans(text))


def redact(text: str) -> str:
    """Return ``text`` with every secret-shaped token replaced by the placeholder.

    Idempotent: redacting already-redacted text is a no-op (the placeholder is not
    itself secret-shaped). Clean text is returned unchanged.
    """
    cleaned = text
    for pattern in _PREFIX_PATTERNS:
        cleaned = pattern.sub(REDACTION_PLACEHOLDER, cleaned)
    cleaned = _ASSIGNMENT_PATTERN.sub(
        lambda m: m.group(0).replace(m.group("val"), REDACTION_PLACEHOLDER), cleaned
    )
    # High-entropy blobs: redact right-to-left so earlier spans keep their offsets.
    for start, end in reversed(_high_entropy_spans(cleaned)):
        cleaned = cleaned[:start] + REDACTION_PLACEHOLDER + cleaned[end:]
    return cleaned


def assert_no_secret(text: str) -> None:
    """Raise :class:`SecretEncounteredError` if ``text`` still contains a secret.

    This is the final gate before a string is written to a proposal/diff/PR/log.
    Callers that can recover should :func:`redact` first; this asserts the invariant.
    """
    if contains_secret(text):
        raise SecretEncounteredError(
            "a secret-shaped token reached an output that must never carry a secret"
        )


__all__ = [
    "REDACTION_PLACEHOLDER",
    "assert_no_secret",
    "contains_secret",
    "redact",
]

File: sdks/typescript/src/onboarding/redact.ts
/**
 * Secret redaction — the TS mirror of Python `valuemaxx.onboarding.redact`.
 *
 * Scrubs secret-shaped tokens (provider key prefixes, secret-named assignments, and
 * high-entropy credential blobs) from any snippet before it leaves the scanner, so a real
 * `.env`-adjacent literal never lands in an onboarding result. The patterns + the entropy
 * threshold come from the shared rules contract, so Python and TS scrub identically (a
 * parity test enforces it). Never executes scanned code.
 */

import type { OnboardingRules } from "./types.js";

/** Bits-per-character Shannon entropy of `token` (0 for empty) — mirrors the Python calc. */
function shannonEntropy(token: string): number {
  if (token.length === 0) return 0;
  const counts = new Map<string, number>();
  for (const ch of token) counts.set(ch, (counts.get(ch) ?? 0) + 1);
  const n = token.length;
  let bits = 0;
  for (const c of counts.values()) {
    const p = c / n;
    bits -= p * Math.log2(p);
  }
  return bits;
}

/**
 * Replace every secret-shaped token in `text` with the redaction placeholder. Order matches
 * Python: prefix patterns first (admin key before generic sk-ant), then the assignment form
 * (redacting only the value), then high-entropy blobs above the entropy threshold.
 */
export function redact(text: string, rules: OnboardingRules): string {
  let out = text;
  const placeholder = rules.redaction_placeholder;

  for (const pattern of rules.redact_prefix_patterns) {
    out = out.replace(new RegExp(pattern, "g"), placeholder);
  }

  // Assignment form: <secret-name> [:=] "<value>" — redact the captured value only. The
  // Python pattern is case-insensitive with a word-boundaried name and a 6+ char value.
  const assignment = new RegExp(
    `\\b(?:${rules.redact_secret_name_alt})\\b\\s*[:=]\\s*("[^"]{6,}"|'[^']{6,}'|[^\\s'"]{6,})`,
    "gi",
  );
  out = out.replace(assignment, (match, value: string) => match.replace(value, placeholder));

  // High-entropy blobs: a long credential-ish run, scrubbed only if its entropy clears the
  // threshold (so ordinary long identifiers/base64-looking-but-low-entropy text survives).
  const blob = new RegExp(rules.redact_high_entropy_pattern, "g");
  out = out.replace(blob, (token) =>
    shannonEntropy(token) >= rules.redact_high_entropy_bits ? placeholder : token,
  );

  return out;
}
